### DISCUSSION TITLE
Formatting Text within a shape.

### DISCUSSION THREAD

**Discussion by gordon-brown** [clientIO/joint#2543] [Q&A]
### Introduction

I want to add margins and/or padding for a shape label. Need to be able to add a margin for text within a shape. There is a lot of documentation on link label, but not so much on shape labels. See old and new shape so that it doesn't overwrite part of the text.

![NewShape](https://github.com/clientIO/joint/assets/32515524/137a347d-7424-4b36-9b40-9d07ddbd2c98)
![OldShape](https://github.com/clientIO/joint/assets/32515524/bd5d47b4-4d62-4713-8c94-5648b888e405)
[Shape.json.txt](https://github.com/clientIO/joint/files/14348203/Shape.json.txt)

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment]
You have to define the margins on a logical level using `calc` expressions. For instance:

```js
'label':  {
  'textAnchor': 'middle',
  'x': 'calc(w/2 + 10)',
  'textWrap': {
    'height': -10,
    'width': -20
  }
}
```

For your particular element, where the rectangle on the left side is roughly 20% of the width of the whole shape, the solution may be as follows:
```js
const textMargin = 5; // keep 5px margin around text
const sidebarRatio = 5; // the sidebar width (on the left) is 20% of the width (100/20)
el.attr({
    label: {
        'x': `calc(w/2 + calc(w / ${2 * sidebarRatio}))`,
        'textWrap': {
            'height': -2 * textMargin,
            'width': `calc(w - calc(w / ${sidebarRatio} + ${2 * textMargin}))`
        }
    }
});
```

Here's a [JSFiddle](https://jsfiddle.net/kumilingus/ojvm20pa/) of this.

**Comment by gordon-brown** [comment]
Thanks for your thorough answer and example. Works as expected.
